h2o-core/src/main/java/water/ClearDKVTask.java

```diff
@@ -0,0 +1,106 @@
+package water;
+
+import hex.Model;
+import water.fvec.Frame;
+import water.fvec.Vec;
+import water.util.Log;
+
+import java.util.*;
+
+/**
+ * Clears keys in all H2O nodes, except for the ones marked as retained.
+ * Only Model and Frame keys are retained. If a key of any other type is provided, it will be removed as well.
+ */
+public class ClearDKVTask extends MRTask<ClearDKVTask> {
+
+  private final Key[] _retainedKeys; // Only model and frame keys
+
+  /**
+   *
+   * @param retainedKeys Keys that are NOT deleted and will remain in DKV. Only Model keys and Frame keys are accepted
+   */
+  public ClearDKVTask(Key[] retainedKeys) {
+    _retainedKeys = retainedKeys;
+  }
+
+  @Override
+  protected void setupLocal() {
+    final Set<Key> retainedKeys = new HashSet<>(_retainedKeys.length);
+    retainedKeys.addAll(Arrays.asList(_retainedKeys));
+    extractNestedKeys(retainedKeys);
+    final Collection<Value> storeKeys = H2O.STORE.values();
+    Futures removalFutures = new Futures();
+    System.out.println(Arrays.toString(retainedKeys.toArray()));
+    for (final Value value : storeKeys) {
+      if (retainedKeys.contains(value._key)) {
+        continue;
+      }
+      if(value.isNull()) continue;
+
+      if (value.isFrame()) {
+        ((Frame) value.get()).retain(removalFutures, retainedKeys);
+      } else if (value.isModel()) {
+        ((Model)value.get()).remove(removalFutures);
+      } 
+    }
+    removalFutures.blockForPending();
+  }
+
+  /**
+   * Iterates through the keys provided by the user, dropping any keys that are not a Model key or a Frame key.
+   * Afterwards, extracts 
+   *
+   * @param retainedKeys A {@link Set} of retained keys to insert the extracted {@link Frame} and {@link Model} keys to.
+   *                     Should contain user-specified keys to retain in order to extract anything.
+   */
+  private static void extractNestedKeys(final Set<Key> retainedKeys) {
+
+    final Iterator<Key> keysIterator = retainedKeys.iterator(); // Traverse keys provided by the user only.
+    while (keysIterator.hasNext()) {
+      final Key key = keysIterator.next();
+      final Value value = Value.STORE_get(key);
```

Comment:
Not all keys have values locally, which is the main point of H2O's existence. This should really be done otherwise - gather all keys to be retained from all nodes in a separate MR Task and then invoke deletion.

This leads into situations where in multi-node (cluster of 5), some keys do get deleted and some tests will fail.